shell/platform/darwin/macos/framework/Source/FlutterKeyEmbedderHandler.mm

```diff
@@ -0,0 +1,605 @@
+ */
+- (void)sendCapsLockTapWithTimestamp:(NSTimeInterval)timestamp
+                            callback:(FlutterKeyHandlerCallback)downCallback;
+
+/**
+ * Send a key event for a modifier key.
+ *
+ * If callback is nil, then the event is synthesized.
+ */
+- (void)sendModifierEventOfType:(BOOL)shouldDown
+                      timestamp:(NSTimeInterval)timestamp
+                        keyCode:(unsigned short)keyCode
+                       callback:(nullable FlutterKeyHandlerCallback)cCallback;
+
+/**
+ * Processes a down event.
+ */
+- (void)handleDownEvent:(nonnull NSEvent*)event
+               callback:(nonnull FlutterKeyHandlerCallback)callback;
+
+/**
+ * Processes an up event.
+ */
+- (void)handleUpEvent:(nonnull NSEvent*)event callback:(nonnull FlutterKeyHandlerCallback)callback;
+
+/**
+ * Processes an up event.
+ */
+- (void)handleCapsLockEvent:(nonnull NSEvent*)event
+                   callback:(nonnull FlutterKeyHandlerCallback)callback;
+
+/**
+ * Processes a flags changed event, where modifier keys are pressed or released.
+ */
+- (void)handleFlagEvent:(nonnull NSEvent*)event
+               callback:(nonnull FlutterKeyHandlerCallback)callback;
+
+/**
+ * Processes the response from a framework.
+ */
+- (void)handleResponse:(BOOL)handled forId:(uint64_t)responseId;
+
+@end
+
+@implementation FlutterKeyEmbedderHandler
+
+- (nonnull instancetype)initWithSendEvent:(FlutterSendEmbedderKeyEvent)sendEvent {
+  self = [super init];
+  if (self != nil) {
+    _sendEvent = sendEvent;
+    _pressingRecords = [NSMutableDictionary dictionary];
+    _pendingResponses = [NSMutableDictionary dictionary];
+    _responseId = 1;
+    _lastModifierFlags = 0;
+    _modifierFlagOfInterestMask = computeModifierFlagOfInterestMask();
+  }
+  return self;
+}
+
+#pragma mark - Private
+
+- (void)synchronizeModifiers:(NSUInteger)currentFlags
+               ignoringFlags:(NSUInteger)ignoringFlags
+                   timestamp:(NSTimeInterval)timestamp {
+  const NSUInteger updatingMask = _modifierFlagOfInterestMask & ~ignoringFlags;
+  const NSUInteger currentFlagsOfInterest = currentFlags & updatingMask;
+  const NSUInteger lastFlagsOfInterest = _lastModifierFlags & updatingMask;
+  NSUInteger flagDifference = currentFlagsOfInterest ^ lastFlagsOfInterest;
+  if (flagDifference & NSEventModifierFlagCapsLock) {
+    [self sendCapsLockTapWithTimestamp:timestamp callback:nil];
+    flagDifference = flagDifference & ~NSEventModifierFlagCapsLock;
+  }
+  while (true) {
+    const NSUInteger currentFlag = lowestSetBit(flagDifference);
+    if (currentFlag == 0) {
+      break;
+    }
+    flagDifference = flagDifference & ~currentFlag;
+    NSNumber* keyCode = [modifierFlagToKeyCode objectForKey:@(currentFlag)];
+    NSAssert(keyCode != nil, @"Invalid modifier flag 0x%lx", currentFlag);
+    if (keyCode == nil) {
+      continue;
+    }
+    BOOL shouldDown = (currentFlagsOfInterest & currentFlag) != 0;
+    [self sendModifierEventOfType:shouldDown
```

Comment:
It can send a lot of messages to the framework. I don't know it's worth it since supporting a variable length array adds non-trivial complexity, especially considering each event already contains a variable length array for `character`.

Reply:
Couldn't it just be a bitmask of the current modifier state? You could convert the bitmask to a platform-independent version here.

Reply:
Actually, I guess it doesn't really matter much.  It's not going to be THAT many events (like maybe four max per event).